biosynseq/visualize.py

```diff
@@ -1,11 +1,632 @@
+
+
+def get_paint_df(fasta_path: Path, embed_path: Path) -> pd.DataFrame:
+    """Collect a set of scalar arrays to paint with.
+
+    Given a path to a fasta file, return a dataframe with information
+    of the GC content and sequence length of each DNA sequence, as well
+    as the molecular weight and isoelectric point of the protein translated
+    from each DNA sequence.
+
+    Parameters
+    ----------
+    fasta_path : Path
+        Path to access fasta sequences.
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe containing information of the GC content, sequence length,
+        molecular weight, and isolelectric point derived from each DNA sequence.
+    """
+    # get DNA sequences
+    dna_seqs = metrics.fasta_to_dna_seqs(fasta_path=fasta_path)
+    # clip DNA sequences to embedding length
+    embed = np.load(embed_path)
+    dna_seqs = dna_seqs[: len(embed)]
+    # translate DNA to protein
+    protein_seqs = metrics.dna_to_protein_seqs(dna_seqs=dna_seqs)
+
+    paint_df = pd.DataFrame(
+        {
+            "GC": metrics.gc_content(dna_seqs),
+            "SequenceLength": metrics.seq_length(dna_seqs),
+            "MolecularWeight": metrics.molecular_weight(protein_seqs),
+            "IsoelectricPoint": metrics.isoelectric_point(protein_seqs),
+        }
+    )
+    return paint_df
+
+
+def run_tsne(embed_data: np.ndarray) -> np.ndarray:
+    """Given 2-dimensional sequence embeddings, return the transformed data using t-SNE.
+
+    Parameters
+    ----------
+    embed_data : np.ndarray
+        Sequence embeddings to be transformed by t-SNE. Must be 2-dimensional.
+
+    Returns
+    -------
+    np.ndarray
+        Transformed embeddings after running t-SNE.
+    """
+    from cuml.manifold import TSNE
+
+    # embed_data must be 2D since rapidsai only supports 2D data
+    model = TSNE(n_components=2, method="barnes_hut")
+    data_proj = model.fit_transform(embed_data)
+    return data_proj
+
+
+def plot_tsne(
+    data_proj: np.ndarray,
+    paint: np.ndarray,
+    paint_name: str,
+    tsne_path: Path,
+    cmap: str = "viridis",
+) -> pd.DataFrame:
+    """Plot t-SNE visualizations for each sequence metric and
+    save the plots as separate images to the specified directory.
+
+    Parameters
+    ----------
+    data_proj : np.ndarray
+        Transformed embeddings after running t-SNE.
+    paint : np.ndarray
+        Dataframe containing information of sequence metrics for each DNA sequence.
+    paint_name : str
+        Name of the sequence metric whose t-SNE visualization will be plotted.
+    tsne_path : Path
+        Path to save t-SNE plots. Must be a directory.
+    cmap : str, optional
+        Colormap to visualize, by default "viridis."
+
+    Returns
+    -------
+    pd.DataFrame
+        Dataframe with plotting values.
+
+    Raises
+    ------
+    ValueError
+        If the given tsne_path is not a directory.
+    """
+    df = pd.DataFrame(
```

Comment:
Is it possible to make a more general plotting function that handles both tsne/umap?